File: app/test/util/filter_sort_test.dart
import 'package:dufs_client/models/dir_entry.dart';
import 'package:dufs_client/models/media_meta.dart';
import 'package:dufs_client/util/filter_sort.dart';
import 'package:flutter_test/flutter_test.dart';

DirEntry _file(String name, {int size = 0, int mtime = 0}) => DirEntry(
      name: name,
      path: '/$name',
      kind: EntryKind.file,
      size: size,
      mtimeMs: mtime,
    );
DirEntry _dir(String name) =>
    DirEntry(name: name, path: '/$name', kind: EntryKind.dir, size: 0);

List<String> _names(List<DirEntry> es) => es.map((e) => e.name).toList();

void main() {
  final meta = <String, MediaMeta>{
    '/short.mp4': const MediaMeta(durationMs: 5000, createdMs: 100, uploadedMs: 200),
    '/long.mp4': const MediaMeta(durationMs: 60000, createdMs: 50, uploadedMs: 400),
  };
  final entries = [
    _dir('Album'),
    _file('photo.jpg', size: 2000, mtime: 30),
    _file('short.mp4', size: 500, mtime: 10),
    _file('long.mp4', size: 9000, mtime: 20),
    _file('notes.txt', size: 10, mtime: 5),
  ];

  group('fuzzyMatch', () {
    test('subsequence, case-insensitive, empty matches', () {
      expect(fuzzyMatch('', 'anything'), isTrue);
      expect(fuzzyMatch('pcj', 'pic.jpg'), isTrue);
      expect(fuzzyMatch('zzz', 'pic.jpg'), isFalse);
    });
  });

  group('categoryOf', () {
    test('maps each kind', () {
      expect(categoryOf(_dir('D')), TypeFilter.folder);
      expect(categoryOf(_file('a.jpg')), TypeFilter.image);
      expect(categoryOf(_file('a.mp4')), TypeFilter.video);
      expect(categoryOf(_file('a.md')), TypeFilter.text);
      expect(categoryOf(_file('a.bin')), TypeFilter.other);
    });
  });

  group('isFilterActive', () {
    test('false for default, true for each field', () {
      expect(isFilterActive(defaultFilter), isFalse);
      expect(isFilterActive(defaultFilter.copyWith(query: 'x')), isTrue);
      expect(isFilterActive(defaultFilter.copyWith(type: TypeFilter.image)),
          isTrue);
      expect(
          isFilterActive(defaultFilter.copyWith(extIncludes: ['jpg'])), isTrue);
      expect(
          isFilterActive(defaultFilter.copyWith(extExcludes: ['jpg'])), isTrue);
      expect(isFilterActive(defaultFilter.copyWith(minSize: 1)), isTrue);
      expect(isFilterActive(defaultFilter.copyWith(maxSize: 1)), isTrue);
      expect(isFilterActive(defaultFilter.copyWith(minDurationMs: 1)), isTrue);
      expect(isFilterActive(defaultFilter.copyWith(maxDurationMs: 1)), isTrue);
      expect(isFilterActive(defaultFilter.copyWith(minPixels: 1)), isTrue);
      expect(isFilterActive(defaultFilter.copyWith(maxPixels: 1)), isTrue);
    });
  });

  group('copyWith', () {
    test('keeps nullable bounds by default, sets them when passed (incl null)',
        () {
      const base = FilterState(
          minSize: 5,
          maxSize: 6,
          minDurationMs: 7,
          maxDurationMs: 8,
          minPixels: 9,
          maxPixels: 10);
      // keep branch: nothing passed → all bounds retained.
      final kept = base.copyWith(query: 'q');
      expect([
        kept.minSize,
        kept.maxSize,
        kept.minDurationMs,
        kept.maxDurationMs,
        kept.minPixels,
        kept.maxPixels,
      ], [5, 6, 7, 8, 9, 10]);
      // set branch: pass explicit values including null.
      final set = base.copyWith(
          minSize: 1,
          maxSize: null,
          minDurationMs: 2,
          maxDurationMs: null,
          minPixels: 3,
          maxPixels: null);
      expect([
        set.minSize,
        set.maxSize,
        set.minDurationMs,
        set.maxDurationMs,
        set.minPixels,
        set.maxPixels,
      ], [1, null, 2, null, 3, null]);
    });
    test('SortState copyWith', () {
      const s = SortState(key: SortKey.size, dir: SortDir.desc);
      expect(s.copyWith(key: SortKey.name).key, SortKey.name);
      expect(s.copyWith(dir: SortDir.asc).dir, SortDir.asc);
      expect(s.copyWith().key, SortKey.size);
    });
  });

  group('applyFilterSort — filtering', () {
    test('type keeps only that category (folders excluded)', () {
      final out = applyFilterSort(
          entries, meta,
          defaultFilter.copyWith(type: TypeFilter.image), defaultSort);
      expect(_names(out), ['photo.jpg']);
    });
    test('folder type keeps only folders', () {
      final out = applyFilterSort(entries, meta,
          defaultFilter.copyWith(type: TypeFilter.folder), defaultSort);
      expect(_names(out), ['Album']);
    });
    test('include extensions (folders pass through)', () {
      final out = applyFilterSort(entries, meta,
          defaultFilter.copyWith(extIncludes: ['mp4']), defaultSort);
      expect(_names(out), ['Album', 'long.mp4', 'short.mp4']);
    });
    test('include multiple extensions', () {
      final out = applyFilterSort(entries, meta,
          defaultFilter.copyWith(extIncludes: ['jpg', 'txt']), defaultSort);
      expect(_names(out), ['Album', 'notes.txt', 'photo.jpg']);
    });
    test('exclude drops matching, keeps extensionless', () {
      final out = applyFilterSort([...entries, _file('README')], meta,
          defaultFilter.copyWith(extExcludes: ['mp4']), defaultSort);
      expect(_names(out), ['Album', 'notes.txt', 'photo.jpg', 'README']);
    });
    test('size range excludes files outside the bounds', () {
      final out = applyFilterSort(entries, meta,
          defaultFilter.copyWith(minSize: 400, maxSize: 3000), defaultSort);
      expect(_names(out), ['Album', 'photo.jpg', 'short.mp4']);
    });
    test('duration range needs an indexed duration in range', () {
      final out = applyFilterSort(
          entries,
          meta,
          defaultFilter.copyWith(minDurationMs: 1000, maxDurationMs: 10000),
          defaultSort);
      expect(_names(out), ['Album', 'short.mp4']);
    });
  });

  group('applyFilterSort — sorting', () {
    test('by name asc/desc, dirs first', () {
      final asc = applyFilterSort(entries, meta, defaultFilter, defaultSort);
      expect(_names(asc),
          ['Album', 'long.mp4', 'notes.txt', 'photo.jpg', 'short.mp4']);
      final desc = applyFilterSort(entries, meta, defaultFilter,
          const SortState(dir: SortDir.desc));
      // Dirs still cluster first even descending.
      expect(desc.first.name, 'Album');
    });
    test('by size, modified, extension, type', () {
      int idx(List<DirEntry> l, String n) => _names(l).indexOf(n);
      final bySize = applyFilterSort(
          entries, meta, defaultFilter, const SortState(key: SortKey.size));
      expect(idx(bySize, 'notes.txt'), lessThan(idx(bySize, 'long.mp4')));
      final byMod = applyFilterSort(
          entries, meta, defaultFilter, const SortState(key: SortKey.modified));
      expect(idx(byMod, 'notes.txt'), lessThan(idx(byMod, 'photo.jpg')));
      final byExt = applyFilterSort(
          entries, meta, defaultFilter,
          const SortState(key: SortKey.extension));
      expect(byExt.first.name, 'Album');
      final byType = applyFilterSort(
          entries, meta, defaultFilter, const SortState(key: SortKey.type));
      expect(byType.first.name, 'Album');
    });
    test('by created, uploaded, duration (from the index)', () {
      int idx(List<DirEntry> l, String n) => _names(l).indexOf(n);
      final byCreated = applyFilterSort(
          entries, meta, defaultFilter, const SortState(key: SortKey.created));
      // long.mp4 createdMs 50 < short.mp4 100.
      expect(idx(byCreated, 'long.mp4'), lessThan(idx(byCreated, 'short.mp4')));
      final byUploaded = applyFilterSort(
          entries, meta, defaultFilter, const SortState(key: SortKey.uploaded));
      expect(
          idx(byUploaded, 'short.mp4'), lessThan(idx(byUploaded, 'long.mp4')));
      final byDur = applyFilterSort(
          entries, meta, defaultFilter, const SortState(key: SortKey.duration));
      expect(idx(byDur, 'short.mp4'), lessThan(idx(byDur, 'long.mp4')));
    });
  });

  group('resolution', () {
    const meta = <String, MediaMeta>{
      '/big.jpg': MediaMeta(width: 4000, height: 3000), // 12 MP
      '/small.jpg': MediaMeta(width: 640, height: 480), // ~0.3 MP
    };
    final entries = [_file('big.jpg'), _file('small.jpg'), _file('none.jpg')];
    int idx(List<DirEntry> es, String n) => es.indexWhere((e) => e.name == n);

    test('min/max pixel filter; unknown dimensions excluded', () {
      expect(
          applyFilterSort(entries, meta,
                  const FilterState(minPixels: 1000000), defaultSort)
              .map((e) => e.name),
          ['big.jpg']);
      expect(
          applyFilterSort(entries, meta,
                  const FilterState(maxPixels: 1000000), defaultSort)
              .map((e) => e.name),
          ['small.jpg']);
    });

    test('sorts by resolution (unknown counts as 0)', () {
      final byRes = applyFilterSort(entries, meta, defaultFilter,
          const SortState(key: SortKey.resolution));
      expect(idx(byRes, 'none.jpg'), lessThan(idx(byRes, 'small.jpg')));
      expect(idx(byRes, 'small.jpg'), lessThan(idx(byRes, 'big.jpg')));
    });
  });
}

File: web/src/lib/filter-sort.test.ts
import { describe, it, expect } from "vitest";
import {
  applyFilterSort,
  categoryOf,
  DEFAULT_FILTER,
  DEFAULT_SORT,
  fuzzyMatch,
  isFilterActive,
  type FilterState,
  type SortState,
} from "./filter-sort.ts";
import type { DirEntry, MetaIndex } from "../api/types.ts";

const dir = (name: string): DirEntry => ({
  name,
  path: `/${name}`,
  kind: "dir",
  size: 0,
  mtimeMs: 0,
});
const file = (name: string, size = 0, mtimeMs = 0): DirEntry => ({
  name,
  path: `/${name}`,
  kind: "file",
  size,
  mtimeMs,
});

const names = (es: readonly DirEntry[]): string[] => es.map((e) => e.name);

const filter = (over: Partial<FilterState> = {}): FilterState => ({
  ...DEFAULT_FILTER,
  ...over,
});
const sort = (over: Partial<SortState> = {}): SortState => ({
  ...DEFAULT_SORT,
  ...over,
});

describe("fuzzyMatch", () => {
  it("empty query matches; subsequence matches; gaps fail", () => {
    expect(fuzzyMatch("", "anything")).toBe(true);
    expect(fuzzyMatch("pic", "my-picture.jpg")).toBe(true);
    expect(fuzzyMatch("PIC", "picture")).toBe(true);
    expect(fuzzyMatch("zzz", "picture")).toBe(false);
  });
});

describe("categoryOf", () => {
  it("classifies all categories", () => {
    expect(categoryOf(dir("d"))).toBe("folder");
    expect(categoryOf(file("a.jpg"))).toBe("image");
    expect(categoryOf(file("a.mp4"))).toBe("video");
    expect(categoryOf(file("a.mp3"))).toBe("audio");
    expect(categoryOf(file("a.pdf"))).toBe("pdf");
    expect(categoryOf(file("a.md"))).toBe("text");
    expect(categoryOf(file("a.bin"))).toBe("other");
  });
});

describe("isFilterActive", () => {
  it("false for default, true for each field", () => {
    expect(isFilterActive(DEFAULT_FILTER)).toBe(false);
    expect(isFilterActive(filter({ query: "x" }))).toBe(true);
    expect(isFilterActive(filter({ type: "image" }))).toBe(true);
    expect(isFilterActive(filter({ extIncludes: ["jpg"] }))).toBe(true);
    expect(isFilterActive(filter({ extExcludes: ["jpg"] }))).toBe(true);
    expect(isFilterActive(filter({ minSize: 1 }))).toBe(true);
    expect(isFilterActive(filter({ maxSize: 1 }))).toBe(true);
    expect(isFilterActive(filter({ minDurationMs: 1 }))).toBe(true);
    expect(isFilterActive(filter({ maxDurationMs: 1 }))).toBe(true);
    expect(isFilterActive(filter({ minPixels: 1 }))).toBe(true);
    expect(isFilterActive(filter({ maxPixels: 1 }))).toBe(true);
  });
});

describe("applyFilterSort — filtering", () => {
  const meta: MetaIndex = {
    "/short.mp4": {
      width: 1920,
      height: 1080,
      durationMs: 5000,
      createdMs: 100,
      uploadedMs: 200,
    },
    "/long.mp4": {
      width: 1280,
      height: 720,
      durationMs: 60000,
      createdMs: 50,
      uploadedMs: 400,
    },
  };
  const entries = [
    dir("Album"),
    file("photo.jpg", 2000, 30),
    file("short.mp4", 500, 10),
    file("long.mp4", 9000, 20),
    file("notes.txt", 10, 5),
  ];

  it("type filter keeps only that category (folders excluded)", () => {
    const out = applyFilterSort(entries, meta, filter({ type: "image" }), sort());
    expect(names(out)).toEqual(["photo.jpg"]);
  });

  it("folder type keeps only folders", () => {
    const out = applyFilterSort(entries, meta, filter({ type: "folder" }), sort());
    expect(names(out)).toEqual(["Album"]);
  });

  it("include-extension filter keeps matching files (folders pass through)", () => {
    const out = applyFilterSort(
      entries,
      meta,
      filter({ extIncludes: ["mp4"] }),
      sort(),
    );
    expect(names(out)).toEqual(["Album", "long.mp4", "short.mp4"]);
  });

  it("include multiple extensions (webp + png style)", () => {
    const out = applyFilterSort(
      entries,
      meta,
      filter({ extIncludes: ["jpg", "txt"] }),
      sort(),
    );
    expect(names(out)).toEqual(["Album", "notes.txt", "photo.jpg"]);
  });

  it("exclude-extension filter drops matching files, keeps extensionless", () => {
    const out = applyFilterSort(
      [...entries, file("README", 1, 1)],
      meta,
      filter({ extExcludes: ["mp4"] }),
      sort(),
    );
    // mp4 files dropped; README (no extension) counts as "not mp4".
    // (name sort is case-insensitive, so README trails photo.jpg.)
    expect(names(out)).toEqual(["Album", "notes.txt", "photo.jpg", "README"]);
  });

  it("size range excludes files outside the bounds", () => {
    const out = applyFilterSort(
      entries,
      meta,
      filter({ minSize: 1000, maxSize: 5000 }),
      sort(),
    );
    expect(names(out)).toEqual(["Album", "photo.jpg"]);
  });

  it("duration range needs the index; unknown durations are excluded", () => {
    const out = applyFilterSort(
      entries,
      meta,
      filter({ minDurationMs: 10000 }),
      sort(),
    );
    expect(names(out)).toEqual(["Album", "long.mp4"]);
    const capped = applyFilterSort(
      entries,
      meta,
      filter({ maxDurationMs: 10000 }),
      sort(),
    );
    expect(names(capped)).toEqual(["Album", "short.mp4"]);
  });

  it("resolution range needs the index; unknown dimensions are excluded", () => {
    // short.mp4 = 1920×1080 (2.07 MP), long.mp4 = 1280×720 (0.92 MP).
    const hi = applyFilterSort(
      entries,
      meta,
      filter({ minPixels: 1_000_000 }),
      sort(),
    );
    expect(names(hi)).toEqual(["Album", "short.mp4"]);
    const lo = applyFilterSort(
      entries,
      meta,
      filter({ maxPixels: 1_000_000 }),
      sort(),
    );
    expect(names(lo)).toEqual(["Album", "long.mp4"]);
  });

  it("fuzzy query narrows folders and files", () => {
    const out = applyFilterSort(entries, meta, filter({ query: "no" }), sort());
    expect(names(out)).toEqual(["notes.txt"]);
  });
});

describe("applyFilterSort — sorting", () => {
  const meta: MetaIndex = {
    "/a.mp4": {
      width: 100,
      height: 100,
      durationMs: 3000,
      createdMs: 300,
      uploadedMs: 10,
    },
    "/b.mp4": {
      width: 200,
      height: 100,
      durationMs: 1000,
      createdMs: 100,
      uploadedMs: 20,
    },
  };
  const entries = [
    file("b.mp4", 20, 200),
    dir("Zed"),
    file("a.mp4", 10, 100),
    dir("Alpha"),
    file("c.txt", 30, 300),
  ];

  it("dirs first, then name asc/desc", () => {
    expect(names(applyFilterSort(entries, meta, DEFAULT_FILTER, sort()))).toEqual(
      ["Alpha", "Zed", "a.mp4", "b.mp4", "c.txt"],
    );
    const desc = applyFilterSort(
      entries,
      meta,
      DEFAULT_FILTER,
      sort({ dir: "desc" }),
    );
    expect(names(desc)).toEqual(["Zed", "Alpha", "c.txt", "b.mp4", "a.mp4"]);
  });

  it("sorts by size, modified, created, uploaded, duration", () => {
    const bySize = applyFilterSort(entries, meta, DEFAULT_FILTER, sort({ key: "size" }));
    expect(names(bySize).slice(2)).toEqual(["a.mp4", "b.mp4", "c.txt"]);
    const byMod = applyFilterSort(entries, meta, DEFAULT_FILTER, sort({ key: "modified" }));
    expect(names(byMod).slice(2)).toEqual(["a.mp4", "b.mp4", "c.txt"]);
    const byCreated = applyFilterSort(entries, meta, DEFAULT_FILTER, sort({ key: "created" }));
    expect(names(byCreated).slice(2)).toEqual(["c.txt", "b.mp4", "a.mp4"]);
    const byUploaded = applyFilterSort(entries, meta, DEFAULT_FILTER, sort({ key: "uploaded" }));
    expect(names(byUploaded).slice(2)).toEqual(["c.txt", "a.mp4", "b.mp4"]);
    const byDur = applyFilterSort(entries, meta, DEFAULT_FILTER, sort({ key: "duration" }));
    expect(names(byDur).slice(2)).toEqual(["c.txt", "b.mp4", "a.mp4"]);
    // a=100×100 (10k px), b=200×100 (20k px), c has no dimensions (0).
    const byRes = applyFilterSort(entries, meta, DEFAULT_FILTER, sort({ key: "resolution" }));
    expect(names(byRes).slice(2)).toEqual(["c.txt", "a.mp4", "b.mp4"]);
  });

  it("sorts by extension and type", () => {
    const byExt = applyFilterSort(entries, meta, DEFAULT_FILTER, sort({ key: "extension" }));
    expect(names(byExt).slice(2)).toEqual(["a.mp4", "b.mp4", "c.txt"]);
    const byType = applyFilterSort(entries, meta, DEFAULT_FILTER, sort({ key: "type" }));
    // files: videos (a,b) before text (c)
    expect(names(byType).slice(2)).toEqual(["a.mp4", "b.mp4", "c.txt"]);
  });
});
